File: bench/binary_trees.lua
local function make_tree(depth)
    if depth == 0 then
        return { left = nil, right = nil, item = 0 }
    end
    local d = depth - 1
    return { left = make_tree(d), right = make_tree(d), item = depth }
end

local function tree_checksum(node)
    if node.left == nil then
        return node.item
    end
    return node.item + tree_checksum(node.left) - tree_checksum(node.right)
end

local MIN_DEPTH = 4
local MAX_DEPTH = 10

local start = os.clock()

local stretch_depth = MAX_DEPTH + 1
local stretch_tree = make_tree(stretch_depth)
local stretch_checksum = tree_checksum(stretch_tree)
print(string.format("stretch tree of depth %d: checksum %d", stretch_depth, stretch_checksum))

local long_lived_tree = make_tree(MAX_DEPTH)

local depth = MIN_DEPTH
while depth <= MAX_DEPTH do
    local iterations = 1
    local p = 0
    while p < depth do
        iterations = iterations * 2
        p = p + 1
    end
    local total_checksum = 0
    local i = 0
    while i < iterations do
        total_checksum = total_checksum + tree_checksum(make_tree(depth))
        i = i + 1
    end
    print(string.format("%d trees of depth %d: checksum %d", iterations, depth, total_checksum))
    depth = depth + 2
end

local long_lived_checksum = tree_checksum(long_lived_tree)
print(string.format("long-lived tree of depth %d: checksum %d", MAX_DEPTH, long_lived_checksum))

local dur = os.clock() - start
print(string.format("binary_trees: %fs", dur))

File: bench/binary_trees.py
import time


class TreeNode:
    __slots__ = ("left", "right", "item")

    def __init__(self, left, right, item):
        self.left = left
        self.right = right
        self.item = item


def make_tree(depth):
    if depth == 0:
        return TreeNode(None, None, 0)
    d = depth - 1
    return TreeNode(make_tree(d), make_tree(d), depth)


def tree_checksum(node):
    if node.left is None:
        return node.item
    return node.item + tree_checksum(node.left) - tree_checksum(node.right)


MIN_DEPTH = 4
MAX_DEPTH = 10

start = time.perf_counter()

stretch_depth = MAX_DEPTH + 1
stretch_tree = make_tree(stretch_depth)
stretch_checksum = tree_checksum(stretch_tree)
print(f"stretch tree of depth {stretch_depth}: checksum {stretch_checksum}")

long_lived_tree = make_tree(MAX_DEPTH)

depth = MIN_DEPTH
while depth <= MAX_DEPTH:
    iterations = 1
    p = 0
    while p < depth:
        iterations *= 2
        p += 1
    total_checksum = 0
    i = 0
    while i < iterations:
        total_checksum += tree_checksum(make_tree(depth))
        i += 1
    print(f"{iterations} trees of depth {depth}: checksum {total_checksum}")
    depth += 2

long_lived_checksum = tree_checksum(long_lived_tree)
print(f"long-lived tree of depth {MAX_DEPTH}: checksum {long_lived_checksum}")

dur = time.perf_counter() - start
print(f"binary_trees: {dur}s")
